### PR TITLE
EHN: Improved thread safety for read_html() GH16928

### LINKED ISSUE #16928: read_html() Thread Safety
#### Code Sample

```python
#!/usr/bin/python3
import pandas
import threading

def fetch_file():
    url = "https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html"
    pandas.read_html(url)

thread1 = threading.Thread(target = fetch_file)
thread2 = threading.Thread(target = fetch_file)

thread1.start()
thread2.start()
```
### Output
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "./pandas_bug.py", line 7, in fetch_file
    pandas.read_html(url)
  File "/usr/lib/python3.6/site-packages/pandas/io/html.py", line 904, in read_html
    keep_default_na=keep_default_na)
  File "/usr/lib/python3.6/site-packages/pandas/io/html.py", line 731, in _parse
    parser = _parser_dispatch(flav)
  File "/usr/lib/python3.6/site-packages/pandas/io/html.py", line 691, in _parser_dispatch
    raise ImportError("lxml not found, please install it")
ImportError: lxml not found, please install it
```
#### Problem description

read_html() doesn't appear to be multi-threading safe. This specific issue seems to be caused by setting `_IMPORTS` in html.py to True too early resulting in the second thread entering `_parser_dispatch` and throwing an exception while the first thread hasn't finished the check.

I have written a potential fix and will open a PR shortly.

#### Expected Output
No exception should be thrown since lxml is installed and the program works fine without multi-threading.
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.11.3-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 36.0.1
Cython: None
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by 3553x**
- [X] closes #16928
 - [x] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [X] whatsnew entry

It failed a total of 6 tests when running `test_fast.sh` but failed none when running `pytest pandas/tests/io/test_html.py`.
It seems that these 6 failed tests are unrelated to my changes since they also occur on the master branch.

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: I'm actually inclined to call this a bug because multi-threading is something that we check for unless explicitly stated otherwise.  Thus, I would move to this to the bugs section.  Also, I would expand your description about the issue was (just a sentence but a longer one would suffice).

Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: use `` :func:`read_html` ``

**Review by jreback — CHANGES_REQUESTED** [review]
can you add the test from the issue

**Comment by 3553x** [comment]
Regarding adding the test.

I have trouble creating a test that reproduces the issue.
The following code snippet works fine when I run `pytest test_html.py::test_importcheck_thread_safety` but as soon as I run `pytest test_html.py` the test always passes. Neither of the threads throw an exception in that case.
```python
def get_html_file():
    filename = os.path.join(DATA_PATH, 'valid_markup.html')
    read_html(filename)


class ErrorThread(threading.Thread):
    def run(self):
        try:
            threading.Thread.run(self)
        except Exception as e:
            self.err = e
        else:
            self.err = None

def test_importcheck_thread_safety():
    t1 = ErrorThread(target = get_html_file)
    t2 = ErrorThread(target = get_html_file)
    t1.start()
    t2.start()
    while(t1.is_alive() or t2.is_alive()):
        pass
    assert None == t1.err == t2.err, "Errors when run in parallel"
```

**Comment by gfyoung** [comment]
> Neither of the threads throw an exception in that case.

I'm confused...isn't that what you want (because it doesn't fail)?

**Comment by 3553x** [comment]
The threads don't throw exceptions regardless of whether I include my fix or not. Meaning the test I wrote is useless and would indicate that everything is fine while actual programs are crashing.

**Comment by gfyoung** [comment]
> The threads don't throw exceptions regardless of whether I include my fix or not.

Ah, okay.  That makes sense.  Does your code from the issue raise an Exception if you incorporate it as a test (just as a starter)?

**Comment by 3553x** [comment]
I'm not sure what you mean by starter. I tried using the code from the issue directly earlier and it didn't work. Apparently exceptions need to be thrown by the main-thread.

**Comment by gfyoung** [comment]
By starter, I meant: does your code from the original issue form a valid test for our code-base (when incorporated into the `tests` directory?  You indeed answered the question.

**Comment by gfyoung** [comment]
@3553x : As I'm not sure at this point why `pytest` is behaving the way it is for you, how about this: you need the test to break and then pass (with your changes) on the build machines.  On a separate branch, add your test, push it, and have Travis and/or Appveyor (either one is sufficient) run and see if it breaks *without your changes*.  Then push your changes to that branch and see if they then pass.

If it does, commit your test and push to this PR.  Otherwise, let us know if it worked or not.  How does that sound to you?

**Comment by 3553x** [comment]
That sounds like something worth trying out. I'll let you know once I'm done.

**Comment by gfyoung** [comment]
One other thing, if this operation is successful, if you could provide us to the links to the failing and passing builds, that way we can see the differential between having and not having your patch.

**Comment by 3553x** [comment]
I tried using travis with my test added on top of the unfixed version. Some builds fail because of linting, however they all seem to pass pytest. I realised that my earlier version might have passed since `_IMPORTS` is set to true after the first call to `read_html()`. So I imported `pandas.io.html` and reloaded it. But this didn't fix the issue. This change was included in the version that was tested by travis.

These are the additions that were made to `test_html.py`:
```python
try:
    from importlib import reload
except ImportError:
    pass
[...]
import pandas.io.html
[...]
class ErrorThread(threading.Thread):
    def run(self):
        try:
            super(ErrorThread, self).run()
        except Exception as e:
            self.err = e
        else:
            self.err = None

@pytest.mark.slow
def test_importcheck_thread_safety():
    reload(pandas.io.html)
    filename = os.path.join(DATA_PATH, 'valid_markup.html')
    helper_thread1 = ErrorThread(target = read_html, args = (filename,))
    helper_thread2 = ErrorThread(target = read_html, args = (filename,))
    helper_thread1.start()
    helper_thread2.start()
    while(helper_thread1.is_alive() or helper_thread2.is_alive()):
        pass
    assert None == helper_thread1.err == helper_thread2.err
```
Output of `pytest test_html.py::test_importcheck_thread_safety`:
```
=========================== test session starts ============================
platform linux -- Python 3.6.1, pytest-3.1.1, py-1.4.34, pluggy-0.4.0
rootdir: /home/fe/code/pandas, inifile: setup.cfg
plugins: xdist-1.16.0
collected 8 items 

test_html.py F

================================= FAILURES =================================
______________________ test_importcheck_thread_safety ______________________

    @pytest.mark.slow
    def test_importcheck_thread_safety():
        reload(pandas.io.html)
        filename = os.path.join(DATA_PATH, 'valid_markup.html')
        helper_thread1 = ErrorThread(target = read_html, args = (filename,))
        helper_thread2 = ErrorThread(target = read_html, args = (filename,))
        helper_thread1.start()
        helper_thread2.start()
        while(helper_thread1.is_alive() or helper_thread2.is_alive()):
            pass
>       assert None == helper_thread1.err == helper_thread2.err
E       AssertionError: assert None == ImportError('lxml not found, please install it',)
E        +  where None = <ErrorThread(Thread-1, stopped 140146488710912)>.err
E        +  and   ImportError('lxml not found, please install it',) = <ErrorThread(Thread-2, stopped 140146480318208)>.err

test_html.py:961: AssertionError
========================= 1 failed in 7.14 seconds =========================
```
Output of `pytest test_html.py`
```
=========================== test session starts ============================
platform linux -- Python 3.6.1, pytest-3.1.1, py-1.4.34, pluggy-0.4.0
rootdir: /home/fe/code/pandas, inifile: setup.cfg
plugins: xdist-1.16.0
collected 74 items 

test_html.py ..........................................................................

======================== 74 passed in 50.77 seconds ========================
```

**Comment by jreback** [comment]
@3553x you can add your test to the PR.

**Comment by jreback** [comment]
pls rebase on master

```
git rebase -i upstream/master
git push yourremote yourbranch -f
```

File: doc/source/whatsnew/v0.21.0.txt
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: importcheck --> "import check"

**Review by gfyoung — APPROVED** [review]

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Could you explain (or at least comment) why we need this?

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Add a newline beneath this (just for readability).

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Add a newline beneath this (just for readability).

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Also, add a newline beneath this (just for readability).

Comment (3553x) [inline pandas/tests/io/test_html.py]: The import check only happens when a variable in `html.py` is set to false, it's initial value. However, the variable will be set to true during the first call to `read_html`. Reloading the module allows us to reinitialise that variable and effectively force an import check.

Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Could you briefly write that as a comment above the `reload` call?

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Let's merge these two `try-except` together since both come from Python 2.x.  Also, make a comment about that `importlib` is from Python 2.x.

File: doc/source/whatsnew/v0.21.0.txt
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: Add the word "where" before "import check"

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Can you reference issue number below the function definition?  Something like "see gh-16928" will suffice.

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: For readability, add a space between "while" and "("

Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Actually, do you need the parentheses?  I don't think you do...

File: doc/source/whatsnew/v0.21.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: say in multiple threads

File: pandas/tests/io/test_html.py
Comment (jreback) [inline pandas/tests/io/test_html.py]: not needed here

File: pandas/tests/io/test_html.py
Comment (jreback) [inline pandas/tests/io/test_html.py]: if you can only test under py3 then use a ``@pytest.mark.skipf(not compat.PY3, reason=.....)`` decorator

**Review by jreback — APPROVED** [review]
lgtm. some comments.

File: pandas/tests/io/test_html.py
Comment (3553x) [inline pandas/tests/io/test_html.py]: I've just tested it with ``python3 `which pytest` test_html.py::test_importcheck_thread_safety`` and  ``python2 `which pytest` test_html.py::test_importcheck_thread_safety`` for both cases (with and without fix). The test seems to work for both versions on my system.

File: pandas/tests/io/test_html.py
Comment (3553x) [inline pandas/tests/io/test_html.py]: Can you elaborate?

Comment (gfyoung) [inline pandas/tests/io/test_html.py]: @3553x : He means to just remove the line.  I thought that line would be useful for clarity?

**Comment by gfyoung** [comment]
@3553x : @jreback has some minor comments that if you could address, we will be able to merge this!

File: pandas/tests/io/test_html.py
Comment (jreback) [inline pandas/tests/io/test_html.py]: if this is not supported in py2, then you should either conditionally import it, or set ``reload=None`` and check it (and skip in a test).

File: pandas/tests/io/test_html.py
Comment (jreback) [inline pandas/tests/io/test_html.py]: so if this works, then remove the comment above

File: pandas/tests/io/test_html.py
Comment (3553x) [inline pandas/tests/io/test_html.py]: `reload` is a built-in in Python2. There is no need to import it unless you are using Python3.

**Comment by gfyoung** [comment]
@3553x : Thanks for updating!

@jreback this LGTM, but I'll leave it to you to check one more time before merging.

File: pandas/tests/io/test_html.py
Comment (jreback) [inline pandas/tests/io/test_html.py]: ok

**Comment by jreback** [comment]
thanks @3553x

**Pull request merged**
